File: java-bridge-matlab/src/main/java/idesyde/matlab/CommunicatingAndTriggeredReactiveWorkloadIRule.java
package idesyde.matlab;

import idesyde.common.CommunicatingAndTriggeredReactiveWorkload;
import idesyde.core.DecisionModel;
import idesyde.core.DesignModel;
import idesyde.core.IdentificationRule;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class CommunicatingAndTriggeredReactiveWorkloadIRule implements IdentificationRule<CommunicatingAndTriggeredReactiveWorkload> {

    @Override
    public Set<CommunicatingAndTriggeredReactiveWorkload> apply(Set<? extends DesignModel> designModels, Set<? extends DecisionModel> decisionModels) {
        var simulinks = designModels.stream().filter(it -> it instanceof SimulinkReactiveDesignModel).map(it -> (SimulinkReactiveDesignModel) it).collect(Collectors.toSet());
        return Set.of(
                new CommunicatingAndTriggeredReactiveWorkload(
                        Stream.concat(simulinks.stream().flatMap(it -> it.processes().stream()), simulinks.stream().flatMap(it -> it.delays().stream())).collect(Collectors.toList()),
                        Stream.concat(simulinks.stream().flatMap(it -> it.processes().stream().map(p -> it.processesSizes().getOrDefault(p, 0L))), simulinks.stream().flatMap(it -> it.delays().stream().map(d -> it.delaysSizes().getOrDefault(d, 0L)))).collect(Collectors.toList()),
                        Stream.concat(simulinks.stream().flatMap(it -> it.processes().stream().map(p -> it.processesOperations().getOrDefault(p, Map.of()))), simulinks.stream().flatMap(it -> it.delays().stream().map(d -> it.delaysOperations().getOrDefault(d, Map.of())))).collect(Collectors.toList()),
                        simulinks.stream().flatMap(it -> it.links().stream()).map(it -> it.toString()).collect(Collectors.toList()),
                        simulinks.stream().flatMap(it -> it.links().stream()).map(it -> it.dataSize()).collect(Collectors.toList()),
                        simulinks.stream().flatMap(it -> it.links().stream()).map(it -> it.src()).collect(Collectors.toList()),
                        simulinks.stream().flatMap(it -> it.links().stream()).map(it -> it.dst()).collect(Collectors.toList()),
                        simulinks.stream().flatMap(it -> it.links().stream()).map(it -> it.dataSize()).collect(Collectors.toList()),
                        simulinks.stream().flatMap(it -> it.sources().stream()).collect(Collectors.toList()),
                        simulinks.stream().flatMap(it -> it.sources().stream().map(s -> it.sourcesPeriods().getOrDefault(s, 0.0))).collect(Collectors.toList()),
                        simulinks.stream().flatMap(it -> it.sources().stream().map(s -> 0.0)).collect(Collectors.toList()),
                        new ArrayList<>(),
                        new ArrayList<>(),
                        new ArrayList<>(),
                        new ArrayList<>(),
                        new ArrayList<>(),
                        new ArrayList<>(),
                        simulinks.stream().flatMap(it -> it.links().stream()).map(it -> it.src()).collect(Collectors.toList()),
                        simulinks.stream().flatMap(it -> it.links().stream()).map(it -> it.dst()).collect(Collectors.toList()),
                        new HashSet<>()
                )
        );
    }
}

File: kotlin-bridge-matlab/src/main/kotlin/idesyde/matlab/CommunicatingAndTriggeredReactiveWorkloadIRule.kt
package idesyde.matlab

import idesyde.common.CommunicatingAndTriggeredReactiveWorkload
import idesyde.core.DecisionModel
import idesyde.core.DesignModel
import idesyde.core.IdentificationRule

class CommunicatingAndTriggeredReactiveWorkloadIRule : IdentificationRule<CommunicatingAndTriggeredReactiveWorkload> {

    override fun usesDecisionModels(): Boolean = false

    override fun partiallyIdentify(
        designModels: Set<DesignModel>,
        decisionModels: Set<DecisionModel>
    ): Set<CommunicatingAndTriggeredReactiveWorkload> {
        val simulinks = designModels.filter { it is SimulinkReactiveDesignModel }.map { it as SimulinkReactiveDesignModel }
        return setOf(
            CommunicatingAndTriggeredReactiveWorkload(
                tasks = simulinks.flatMap { it.processes } + simulinks.flatMap { it.delays },
                taskSizes =
                simulinks.flatMap { s -> s.processes.map { s.processesSizes[it] ?: 0L } } +
                        simulinks.flatMap { s -> s.delays.map { s.delaysSizes[it] ?: 0L} },
                taskComputationalNeeds =
                simulinks.flatMap { s -> s.processes.map { s.processesOperations[it] ?: mapOf() } } +
                        simulinks.flatMap { s -> s.delays.map { s.delaysOperations[it] ?: mapOf() } },
                dataChannels =
                simulinks.flatMap { it.links }.map { it.toString() },
                dataChannelSizes = simulinks.flatMap { it.links }.map { it.dataSize },
                dataGraphSrc = simulinks.flatMap { it.links }.map { it.src },
                dataGraphDst = simulinks.flatMap { it.links }.map { it.dst },
                dataGraphMessageSize = simulinks.flatMap { it.links }.map { it.dataSize },
                periodicSources = simulinks.flatMap { it.sources },
                periods  = simulinks.flatMap { s -> s.sources.map { s.sourcesPeriods[it] ?: 0.0} },
                offsets = simulinks.flatMap { s -> s.sources.map { 0.0 } },
                upsamples = listOf(),
                upsampleRepetitiveHolds = listOf(),
                upsampleInitialHolds = listOf(),
                downsamples = listOf(),
                downampleRepetitiveSkips = listOf(),
                downampleInitialSkips = listOf(),
                triggerGraphSrc = simulinks.flatMap { it.links }.map { it.src },
                triggerGraphDst = simulinks.flatMap { it.links }.map { it.dst },
                hasORTriggerSemantics = setOf()
            )
        )
    }
}